### PR TITLE
SparK Implementation

### PR DESCRIPTION
**Pull request by gabrielfruet**
Addresses #1462 

Continues #1750

Inspired by: https://github.com/keyu-tian/SparK

https://arxiv.org/pdf/2301.03580

### CODE REVIEW

**Comment by guarin** [comment]
Hi, both are maintained :) Just didn't have a lot of time to work on the lightly repo recently. This repo is for SSL research whereas LightlyTrain is more targeted towards production/industry use-cases and also covers fine-tuning etc.

New methods are the most interesting contributions for us. Especially recent methods like:
* JEPAs
* Pixio

There are also some opportunities to improve the benchmarking code. Let me know if you are interested in this.

Thanks a lot for the interest and contributions!

File: examples/pytorch_lightning/spark.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        )
+        downsample_ratio = get_downsample_ratio_from_timm_model(backbone)
+        enc_feat_map_chs = get_enc_feat_map_chs_from_timm_model(backbone)
+        self.sparse_encoder = dense_model_to_sparse(backbone, sbn=sbn, verbose=True)
+        self.fmap_h = input_size // downsample_ratio
+        self.fmap_w = input_size // downsample_ratio
+        self.dense_decoder = UNetDecoder(
+            downsample_ratio,
+            width=enc_feat_map_chs[-1],
+        )
```
Comment (liopeer) [inline examples/pytorch_lightning/spark.py]: Prefer to use kwargs for any calls with more than a single argument, makes your code easier to read.

File: lightly/loss/sparse_spark.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        """
+        if active_mask.ndim != 4:
+            raise ValueError(
+                "active_mask must be non-flattened with shape (B, 1, f, f)"
+            )
+
+        mean = inp_patches.mean(dim=-1, keepdim=True)
+        var = (inp_patches.var(dim=-1, keepdim=True) + self.eps) ** 0.5
+
+        inp_norm = (inp_patches - mean) / var
```
Comment (liopeer) [inline lightly/loss/sparse_spark.py]:
```suggestion
        inp_normalized = (inp_patches - mean) / var
```
Since it's not the "norm" of the input, but the "normalized" input.

File: lightly/loss/sparse_spark.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+        l2_loss = ((rec_patches - inp_norm) ** 2).mean(dim=2)
+
+        non_active = active_mask.logical_not().int().view(active_mask.shape[0], -1)
+
+        recon_loss = l2_loss.mul_(non_active).sum() / (non_active.sum() + 1e-8)
```
Comment (liopeer) [inline lightly/loss/sparse_spark.py]: I would usually refrain from using in-place operations like `mul_`. They might cause issues with things like `torch.compile` or autograd, i.e. wherever the computational graph has to be constructed.

File: lightly/models/modules/sparse_spark.py [hunk 4]
```diff
@@ -0,0 +1,808 @@
...
+    def forward(self, inp_bchw: torch.Tensor) -> SparKMaskingOutput:
+        """Generate hierarchical masks for the input.
+
+        Creates masks at multiple scales from the patch level up to full input resolution.
+
+        Args:
+            inp_bchw: Input image tensor of shape (B, C, H, W).
+
+        Returns:
+            SparKMaskingOuptut containing:
```
Comment (liopeer) [inline lightly/models/modules/sparse_spark.py]:
```suggestion
            SparKMaskingOutput containing:
```

**Comment by gabrielfruet** [comment]
Hi @liopeer, that's fine! I personally understand.

I will address some issues in this PR, that i acknowledged through the modifications that Igor made to my LeJEPA PR.

But would still be nice to have some direction here! Just to guarantee we're in the same page.

**Review by liopeer — COMMENTED** [review]
I left some comments that you can already address. For an updated styleguide you can check #1895 where I added a few more instructions. I will look on our side into:
- do we want/need to activate linting for `examples`
- could a context manager for the global tensor be an option?

File: lightly/models/modules/sparse_spark.py [hunk 5]
```diff
@@ -0,0 +1,808 @@
+# Copyright (c) 2023 Keyu Tian
```
[the remark this answers is not in the document]
Comment (gabrielfruet) [reply lightly/models/modules/sparse_spark.py]: https://github.com/keyu-tian/SparK

This is the original repo. I joined some files into this single file and after that i started refactoring the classes and functions to adhere to the Lightly philosophy of providing the modules, not the full implementation.

**Comment by gabrielfruet** [comment]
Hi @liopeer , when possible, can you approve these workflows? I think i fixed most of the issues. Also, the 3.7 tests are still relevant?

Also if you have any other suggestion for the code, let me know!

**Comment by gabrielfruet** [comment]
1. Tests are green
2. Formatted the code
3. Generated notebooks

Quick question: One of my modules depends on `timm`. Should i port the necessary classes to our repo or should i declare it as a dependency?

```python
from timm.models.layers import DropPath, trunc_normal_
```

**Comment by IgorSusmelj** [comment]
> 1. Tests are green
> 2. Formatted the code
> 3. Generated notebooks
> 
> Quick question: One of my modules depends on `timm`. Should i port the necessary classes to our repo or should i declare it as a dependency?
> 
> ```python
> from timm.models.layers import DropPath, trunc_normal_
> ```

If timm is not yet a dependency we should copy over the files. You can add a note that you copied them from timm.

**Comment by gabrielfruet** [comment]
There are two options here. I was taking a look and we also have this pattern for timm's ViT to skip tests on minimal CI

https://github.com/lightly-ai/lightly/blob/904c50a7cc8bff9dbc0863a99bc71c84422322a8/lightly/utils/dependency.py#L27-L44

We already have a test that depends on `DropPath` and is being just skipped if timm is not available. Since we are already using this pattern, i'll stick to it.

https://github.com/lightly-ai/lightly/blob/904c50a7cc8bff9dbc0863a99bc71c84422322a8/tests/models/test_ModelUtils.py#L923-L938

**Comment by gabrielfruet** [comment]
I tested with py37 on my environment and the workflows went green. I'm just worried about the test coverage. I'll take a look into it when possible.

**Comment by liopeer** [comment]
Don't worry too much about the test coverage. :) I can review again mid next week.

**Comment by gabrielfruet** [comment]
1. Was missing `trunc_normal_` import on one of the files
2. Check 3.7 was failling due to pre 3.8 type hints.

Iterating a bit through the workflows, still getting used to it :)

**Comment by gabrielfruet** [comment]
Failed because i forgot to run `make format`. I opened an issue to discuss a solution for this #1908 using pre-commit.

File: examples/notebooks/pytorch_lightning/spark.ipynb [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "10",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "dataset = torchvision.datasets.Caltech101(\n",
```
Comment (liopeer) [inline examples/notebooks/pytorch_lightning/spark.ipynb:235]: The download of Caltech101 currently does not seem to work for me. If it worked for you, ignore this comment – might open an issue with torchvision after checking if it has something to do with my local setup.

File: tests/models/modules/test_sparse_spark.py [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+def test__get_active_ex_or_ii_dont_shrink_mask() -> None:
+    H, W = 4, 4
+    with sparse_spark.sparse_layer_context(torch.ones(1, 1, 32, 32)):
+        with pytest.raises(AssertionError):
+            sparse_spark._get_active_ex_or_ii(H=H, W=W, returning_active_ex=False)
+
+
+def test__get_active_ex_or_ii_raise_on_non_active_mask() -> None:
+    H, W = 32, 32
+    with pytest.raises(RuntimeError):
```
Comment (liopeer) [inline tests/models/modules/test_sparse_spark.py:53]: Best to include a `match` parameter with `pytest.raises`. That also serves as documentation on what the test is supposed to check.

**Review by liopeer — COMMENTED** [review]
Overall I think this is mostly done. I will spend some time on Monday to rename a few variables and do smaller clean-ups – some variables are not very descriptive and I would like to change that before we make this public-facing. Thanks a lot for the great contribution!

File: examples/notebooks/pytorch_lightning/spark.ipynb (hunk 6 above)
Comment (gabrielfruet) [reply examples/notebooks/pytorch_lightning/spark.ipynb:235]: I know that `torchvision` is deprecating some datasets. So, some links are broken https://github.com/pytorch/vision/pull/9335

File: tests/models/modules/test_sparse_spark.py (hunk 7 above)
Comment (gabrielfruet) [reply tests/models/modules/test_sparse_spark.py:53]: I'll include!
